lib/assertions.js

```diff
@@ -344,9 +344,100 @@ module.exports = expect => {
           this.outdentLines();
         });
       }
-      propertyNames.forEach(propertyName => {
-        expect(subject, '[not] to have [own] property', String(propertyName));
-      });
+
+      if (expect.flags.only) {
+        if (expect.flags.not) {
+          expect.errorMode = 'bubble';
+          expect.fail(
+            'The "not" flag cannot be used together with "to only have properties".'
+          );
+        }
+        if (expect.flags.own) {
+          expect.errorMode = 'bubble';
+          expect.fail(
+            'The "own" flag cannot be used together with "to only have properties".'
+          );
+        }
+        const subjectType = expect.subjectType;
+        const subjectKeys = subjectType.getKeys(subject).filter(
+          key =>
+            utils.numericalRegExp.test(key) ||
+            typeof key === 'symbol' ||
+            // include only those keys whose value is not undefined
+            typeof subjectType.valueForKey(subject, key) !== 'undefined'
+        );
+
+        expect.withError(
+          () => {
+            expect(subjectKeys.length === propertyNames.length, 'to be true');
+            // now catch differing property names
+            expect(subjectKeys, 'to equal', propertyNames);
```

Comment:
This will only work if you sort the arrays being compared.

```js
expect({ foo: 123, bar: 456 }, 'to only have properties', ['bar', 'foo']);
```

Fails with:

```js
expected { foo: 123, bar: 456 } to only have properties [ 'bar', 'foo' ]

{
  foo: 123,
  bar: 456
}
```

Reply:
Hah yes it would, mea culpa 🙋‍♂️